Added support for BrowserSync

tkdave opened the pull request:
Updated the start task to add support for BrowserSync.
App contents are reloaded when app.js or html is changed.
CSS changes are streamed to the client.

szwacz commented on tasks/utils.js:
```diff
@@ -50,3 +51,39 @@ module.exports.spawnablePath = function (path) {
+    var err;
+    try {
+        bs.init(opts, function (){
+            ready( err ? null : bs);
```
This is rather strange use of error and I couldn't image this code actually works ;)
if `bs.init` throws how callback function can be invoked at all?

tkdave replied:
I know this is unconventional, but it was the only way I could figure out how to get it to fail gracefully, because, for some reason, the callback IS being called regardless of an error being thrown. So, I am catching it below and storing the message in an in-scope 'err' variable to output it to the log before we continue (albeit without browserSync running). Try changing the port to 358299999. You will see an error message. BrowserSync will not start. But, the app will start anyway.

tkdave replied:
I know this is unconventional, but it was the only way I could figure out how to get it to fail gracefully, because, for some reason, the callback IS being called regardless of an error being thrown. So, I am catching it below and storing the message in an in-scope 'err' variable to output it to the log before we continue (albeit without browserSync running). Try changing the port to 358299999. You will see an error message. BrowserSync will not start. But, the app will start anyway.

szwacz commented on tasks/start.js:
```diff
@@ -3,13 +3,40 @@
-        // User closed the app. Kill the host process.
-        process.exit();
+
+    utils.initBrowserSync({port: 3002}, function(bs) {
```
I believe there is no sense in splitting this code into two places (here and utils file).
If you declare everything in `start.js` the code will also become a little simpler.

tkdave replied:
I originally had everything in start.js, but it seemed a bit clunky. I leaned toward moving things to a separate file (in this case the existing utils.js). If you don't mind a bunch more code in start.js, I'm happy to move those functions there and simplify a bit.

szwacz replied:
Yes, please. It makes more sense from separation of concernes standpoint. You can also create totally separate file for browserSync and just require it in start. Might be even better this way.

szwacz commented on tasks/utils.js:
```diff
@@ -50,3 +51,39 @@ module.exports.spawnablePath = function (path) {
+
+// Setup BrowserSync and catch errors gracefully
+module.exports.initBrowserSync = function (bs_opts, ready) {
+    var bs = require("browser-sync").create();
```
Please move `require()` declaration to the top of file.

tkdave replied:
The idea here is to only load those libs if we are using them.

szwacz replied:
When the code will be moved to separate file this problem disappears.

szwacz commented:
Hi @tkdave, 

Pretty slick solution. I only left few general comments about code organisation. If you can apply them, or say why what I said isn't good idea we can merge it.

tkdave commented:
To give credit where it's due, much of this was inspired by the example here:
https://github.com/papaeye/electron-browser-sync-example

tkdave commented:
I will take one more pass at code organization later today, and you can merge at your leisure. This is all working great for me in my app.

gbaumgart commented:
-1 for for adding more useless modules to the template

gbaumgart commented:
@tkdave I mean, you only need to point to your real web-server with all the dev-stuff inside you need and pass the electron instance along as max. There is zero need to add another hundred deps for something which doesn't belong here anyway. Same goes actually for the gulp-watch and and gulp less btw. Its plain actually awful to strip that stuff off. If that would actually happen I'd be happy to contribute about real problems with this template. sorry, just saying. makes zero sense.

tkdave commented:
@gbaumgart, I don't fully understand your followup comment about a real web server. What do you mean by 'pass the electron instance along as max'?

To your original point, BrowserSync is far from useless. If it were just reloading the page, I might understand your sentiment, but this is a huge productivity boost, especially when working on CSS.

I also disagree about the inclusion of gulp-watch and less. I evaluated a variety of electron boilerplates, and this one really hit the sweet spot of providing essential build and packaging tools while remaining fairly agnostic about JavaScript framework choice. I also found it quite customizable. In my case, I replaced less with sass, and the switch was painless. The only other thing I was really missing was BrowserSync, so I added it. Then, as I noticed people asking for this feature (2 of the 6 open issues are requests for LiveReload functionality), I spent part of my Saturday cleaning up my changes and submitting this pull request.

I'm sure there is room on github for a lighter, slimmed-down version of this boilerplate that doesn't include so many gulp plugins. Maybe you should make one. I'd be happy to review it for you.

gbaumgart commented:
small template: yes, i am actually on it and I was actually thinking to send it as PR to here but I just realize this goes now a little out of scope. 

By web-server I mean: Usually when your project really worth it to package it something like with Electron, it means already you've developed it already outside of this template, on a web-server. And you usually have all all the tools like browser-sync there but not in the export template. 
So, as long you point this electron app to your project url, there is no need to repeat the same functionality inside the template. All you need to do is getting the "Electron" instance for Browser-Window, etc.. inside your actual project and then you go on with coding/testing. 

What you're are trying to do is, to abuse this export template into full development stack, with all the neat stuff like CSS, JS post-processors, etc.. 

At the end, we only want to write: 

electron-template --appDir=... --configDir=... --url=.... start 
electron-template --appDir=... --configDir=... --url=.... release
....

well, never mind, go on as you like but I am just saying, such coding tools belong in your project, not here.

tkdave commented:
browser-sync is a rather heavy dependency. We could make it optional or even leave it out of the package.json.  If dependency is met, we init the browser-sync stuff. Otherwise, we just continue without it and leave a line in the log...

`Note: BrowserSync is available. "npm install browser-sync" to enable it.`

Thoughts?

gbaumgart commented:
I use a lot [requireg](https://github.com/gbaumgart/requireg). In my fork of this electron-template I am pulling the electron binaries with that. So that I can build the binaries in virtual-machines on a network-drive. 

In that you pull it via requireg('browser-sync'). Adding a script to install that optional stuff in global shouldn't be a problem, way better than the other way around.
g

szwacz commented:
I can +1 both sides of the argument here (However strange it sounds).

@gbaumgart I share your viewpoint about not bloating this project, and you're right that we're close to that line. That being said I think there is a place for live reload here.

Just a loose thoughts....

0. Whatever change I'll intoduce to this codebase, at least one user will be upset and at least one user happy :)
1. Actually this whole project is just a bunch of tools wired together. The know-how about establishing Electron app.
2. As long as removing stuff which you don't want to have is easy there is no problem in having a little more stuff.
3. About LESS - using CSS preprocessors is standard today, and probably if it won't be here 90% of people will add some anyway. LESS and SASS are the prominent two and switching from one to other is changing actually few lines of code and extension of one file.
4. And about live reload - similarly as LESS, probably more than 50% of people will add it anyway, so it's fine as long as it's also easy to strip it from the code.
5. Deleting whole files and some lines in `package.json` is the easiest form of removing functionality I can image.

Must go now. Will return to this debate.

gbaumgart commented:
@szwacz, @tkdave
yeah, don't bother, if you think its really needed in an exporter, do as you like. I gonna go on with my template. I had to kick all that bloat stuff out and also switch to grunt (better debugging and plugins). Maybe it could serve as template some day for this package here. May be you split the toy stuff from the actual thing one day.... lets see

carlosperate commented:
> 6 - Deleting whole files and some lines in package.json is the easiest form of removing functionality I can image.

While I do agree that I'd prefer this boilerplate to stay slim, this is definitely the best approach for this. If a feature has all its logic in a single source file, and then a line or two to call it somewhere else, then it really simplifies the process of removing such feature AND maintaining the boilerplate updated without it.

szwacz commented:
I wasn't fully happy with integration of browser sync in this PR, but merged it in good faith I can tweak it so I can like it. Unfortunately...
- I felt on my skin how heavy dependency BrowserSync is. This is rather overkill for such simple task as reloading and CSS injection.
- BrowserSync has been hacked a little to work with this boilerplate at all (calling directly internal method of BS: `connectUtils.clientScript(options)`) this is not part of official API and because of that can break in the future.
- BrowserSync duplicates file watchers which gulp watch does as well (too un-optimized solution). Although was able to resolve this problem by integrating browsersync and watch together.

Next looked into alternative solutions, and found very interesting [electron-connect](https://github.com/Quramy/electron-connect). This solution is nice mainly because can restart browser and renderer process (but doesn't have CSS re-injection). Also integration of electron-connect is rather trivial so not much gain in providing it here.

So... I reverted the merge, and it looks like this project will go without live reload for now, simply because there is no full-featured solution to pick.

Sorry folks, I hope I sold my point of view.

carlosperate commented:
I might be biased about my point of view on this specific feature, but it does seem like you are taking the best course of action given the circumstances.
